# Sales Analysis 


## Libraries

In [91]:
import pandas as pd
import numpy as np
import plotly.express as px
from scipy import stats as st
import plotly.graph_objects as go
from plotly.subplots import make_subplots

print('All libraries are ready to use.')

All libraries are ready to use.


## Data overview and first look

In [92]:
data = pd.read_csv('sales_data_sample.csv',encoding='ISO-8859-1')
print('Data is ready.')

Data is ready.


In [93]:
column = data.shape[1]
rows = data.shape[0]

print(f'There are {column} columns and {rows} rows in the Sales dataset.')

There are 25 columns and 2823 rows in the Sales dataset.


In [94]:
data.columns.to_list()

['ORDERNUMBER',
 'QUANTITYORDERED',
 'PRICEEACH',
 'ORDERLINENUMBER',
 'SALES',
 'ORDERDATE',
 'STATUS',
 'QTR_ID',
 'MONTH_ID',
 'YEAR_ID',
 'PRODUCTLINE',
 'MSRP',
 'PRODUCTCODE',
 'CUSTOMERNAME',
 'PHONE',
 'ADDRESSLINE1',
 'ADDRESSLINE2',
 'CITY',
 'STATE',
 'POSTALCODE',
 'COUNTRY',
 'TERRITORY',
 'CONTACTLASTNAME',
 'CONTACTFIRSTNAME',
 'DEALSIZE']

In [95]:
data.dtypes

ORDERNUMBER           int64
QUANTITYORDERED       int64
PRICEEACH           float64
ORDERLINENUMBER       int64
SALES               float64
ORDERDATE               str
STATUS                  str
QTR_ID                int64
MONTH_ID              int64
YEAR_ID               int64
PRODUCTLINE             str
MSRP                  int64
PRODUCTCODE             str
CUSTOMERNAME            str
PHONE                   str
ADDRESSLINE1            str
ADDRESSLINE2            str
CITY                    str
STATE                   str
POSTALCODE              str
COUNTRY                 str
TERRITORY               str
CONTACTLASTNAME         str
CONTACTFIRSTNAME        str
DEALSIZE                str
dtype: object

In [96]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ORDERNUMBER       2823 non-null   int64  
 1   QUANTITYORDERED   2823 non-null   int64  
 2   PRICEEACH         2823 non-null   float64
 3   ORDERLINENUMBER   2823 non-null   int64  
 4   SALES             2823 non-null   float64
 5   ORDERDATE         2823 non-null   str    
 6   STATUS            2823 non-null   str    
 7   QTR_ID            2823 non-null   int64  
 8   MONTH_ID          2823 non-null   int64  
 9   YEAR_ID           2823 non-null   int64  
 10  PRODUCTLINE       2823 non-null   str    
 11  MSRP              2823 non-null   int64  
 12  PRODUCTCODE       2823 non-null   str    
 13  CUSTOMERNAME      2823 non-null   str    
 14  PHONE             2823 non-null   str    
 15  ADDRESSLINE1      2823 non-null   str    
 16  ADDRESSLINE2      302 non-null    str    
 17  CITY  

In [97]:
data.head(5)

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,QTR_ID,MONTH_ID,YEAR_ID,...,ADDRESSLINE1,ADDRESSLINE2,CITY,STATE,POSTALCODE,COUNTRY,TERRITORY,CONTACTLASTNAME,CONTACTFIRSTNAME,DEALSIZE
0,10107,30,95.70,2,2871.00,2/24/2003 0:00,Shipped,1,2,2003,...,897 Long Airport Avenue,NaN,NYC,NY,10022,USA,NaN,Yu,Kwai,Small
1,10121,34,81.35,5,2765.90,5/7/2003 0:00,Shipped,2,5,2003,...,59 rue de l'Abbaye,NaN,Reims,NaN,51100,France,EMEA,Henriot,Paul,Small
2,10134,41,94.74,2,3884.34,7/1/2003 0:00,Shipped,3,7,2003,...,27 rue du Colonel Pierre Avia,NaN,Paris,NaN,75508,France,EMEA,Da Cunha,Daniel,Medium
3,10145,45,83.26,6,3746.70,8/25/2003 0:00,Shipped,3,8,2003,...,78934 Hillside Dr.,NaN,Pasadena,CA,90003,USA,NaN,Young,Julie,Medium
4,10159,49,100.00,14,5205.27,10/10/2003 0:00,Shipped,4,10,2003,...,7734 Strong St.,NaN,San Francisco,CA,NaN,USA,NaN,Brown,Julie,Medium


In [98]:
data.sample(5)

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,QTR_ID,MONTH_ID,YEAR_ID,...,ADDRESSLINE1,ADDRESSLINE2,CITY,STATE,POSTALCODE,COUNTRY,TERRITORY,CONTACTLASTNAME,CONTACTFIRSTNAME,DEALSIZE
354,10253,25,90.17,14,2254.25,6/1/2004 0:00,Cancelled,2,6,2004,...,Berkeley Gardens 12 Brewery,NaN,Liverpool,NaN,WX1 6LT,UK,EMEA,Devon,Elizabeth,Small
2791,10361,23,95.20,12,2189.60,12/17/2004 0:00,Shipped,4,12,2004,...,"Monitor Money Building, 815 Pacific Hwy",Level 6,Chatswood,NSW,2067,Australia,APAC,Huxley,Adrian,Small
2788,10329,44,86.13,8,3789.72,11/15/2004 0:00,Shipped,4,11,2004,...,897 Long Airport Avenue,NaN,NYC,NY,10022,USA,NaN,Yu,Kwai,Medium
833,10139,41,100.00,8,7956.46,7/16/2003 0:00,Shipped,3,7,2003,...,"Monitor Money Building, 815 Pacific Hwy",Level 6,Chatswood,NSW,2067,Australia,APAC,Huxley,Adrian,Large
2648,10223,29,100.00,8,3199.86,2/20/2004 0:00,Shipped,1,2,2004,...,636 St Kilda Road,Level 3,Melbourne,Victoria,3004,Australia,APAC,Ferguson,Peter,Medium


In [99]:
data.describe()

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,QTR_ID,MONTH_ID,YEAR_ID,MSRP
count,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.00000,2823.000000
mean,10258.725115,35.092809,83.658544,6.466171,3553.889072,2.717676,7.092455,2003.81509,100.715551
std,92.085478,9.741443,20.174277,4.225841,1841.865106,1.203878,3.656633,0.69967,40.187912
min,10100.000000,6.000000,26.880000,1.000000,482.130000,1.000000,1.000000,2003.00000,33.000000
25%,10180.000000,27.000000,68.860000,3.000000,2203.430000,2.000000,4.000000,2003.00000,68.000000
50%,10262.000000,35.000000,95.700000,6.000000,3184.800000,3.000000,8.000000,2004.00000,99.000000
75%,10333.500000,43.000000,100.000000,9.000000,4508.000000,4.000000,11.000000,2004.00000,124.000000
max,10425.000000,97.000000,100.000000,18.000000,14082.800000,4.000000,12.000000,2005.00000,214.000000


In [100]:
data.nunique()

ORDERNUMBER          307
QUANTITYORDERED       58
PRICEEACH           1016
ORDERLINENUMBER       18
SALES               2763
ORDERDATE            252
STATUS                 6
QTR_ID                 4
MONTH_ID              12
YEAR_ID                3
PRODUCTLINE            7
MSRP                  80
PRODUCTCODE          109
CUSTOMERNAME          92
PHONE                 91
ADDRESSLINE1          92
ADDRESSLINE2           9
CITY                  73
STATE                 16
POSTALCODE            73
COUNTRY               19
TERRITORY              3
CONTACTLASTNAME       77
CONTACTFIRSTNAME      72
DEALSIZE               3
dtype: int64

In [101]:
data.memory_usage(deep= True)

Index                 132
ORDERNUMBER         22584
QUANTITYORDERED     22584
PRICEEACH           22584
ORDERLINENUMBER     22584
SALES               22584
ORDERDATE           62232
STATUS              42649
QTR_ID              22584
MONTH_ID            22584
YEAR_ID             22584
PRODUCTLINE         53397
MSRP                22584
PRODUCTCODE         45481
CUSTOMERNAME        81790
PHONE               55434
ADDRESSLINE1        77522
ADDRESSLINE2        25439
CITY                44471
STATE               26821
POSTALCODE          37258
COUNTRY             36825
TERRITORY           30054
CONTACTLASTNAME     40768
CONTACTFIRSTNAME    38617
DEALSIZE            38083
dtype: int64

## Cleaning

### Overall

In [102]:
data.columns.to_list()

['ORDERNUMBER',
 'QUANTITYORDERED',
 'PRICEEACH',
 'ORDERLINENUMBER',
 'SALES',
 'ORDERDATE',
 'STATUS',
 'QTR_ID',
 'MONTH_ID',
 'YEAR_ID',
 'PRODUCTLINE',
 'MSRP',
 'PRODUCTCODE',
 'CUSTOMERNAME',
 'PHONE',
 'ADDRESSLINE1',
 'ADDRESSLINE2',
 'CITY',
 'STATE',
 'POSTALCODE',
 'COUNTRY',
 'TERRITORY',
 'CONTACTLASTNAME',
 'CONTACTFIRSTNAME',
 'DEALSIZE']

In [103]:
data.columns = data.columns.str.lower()

In [104]:
data.columns.to_list()

['ordernumber',
 'quantityordered',
 'priceeach',
 'orderlinenumber',
 'sales',
 'orderdate',
 'status',
 'qtr_id',
 'month_id',
 'year_id',
 'productline',
 'msrp',
 'productcode',
 'customername',
 'phone',
 'addressline1',
 'addressline2',
 'city',
 'state',
 'postalcode',
 'country',
 'territory',
 'contactlastname',
 'contactfirstname',
 'dealsize']

In [105]:
missing_values = data.isnull().sum()
missing_pct = (missing_values/len(data)*100).round(1)

missing_summary = pd.DataFrame({'missing_count':missing_values , 'missing_rate': missing_pct })

missing_summary[missing_summary['missing_count']>0]

,missing_count,missing_rate
addressline2,2521,89.3
state,1486,52.6
postalcode,76,2.7
territory,1074,38.0


**This clearly shows that there are 4 columns `Addressline`,`state`,`postalcode`,`territory` have some missing values**

In [106]:
columns_to_fill = ['addressline2','state','postalcode','territory']

data[columns_to_fill] = data[columns_to_fill].fillna('unknown')

In [107]:
data[columns_to_fill].isnull().sum()

addressline2    0
state           0
postalcode      0
territory       0
dtype: int64

in the cell above we can clearly see that all of the null values of 4 columns filled with `unknown`

In [108]:
data['orderdate'].dtypes

<StringDtype(na_value=nan)>

From above orderdate is string 

In [109]:
data['orderdate'] = pd.to_datetime(data['orderdate'])
print(data['orderdate'].dtype)

datetime64[us]


### Duplicates

In [110]:
dup_counts = data.duplicated().sum()
print(f'Total duplicates count: {dup_counts}')

Total duplicates count: 0


This show that there is no duplicated totalt rows

## Statistics

In [111]:
data.describe()

,ordernumber,quantityordered,priceeach,orderlinenumber,sales,orderdate,qtr_id,month_id,year_id,msrp
count,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823,2823.000000,2823.000000,2823.00000,2823.000000
mean,10258.725115,35.092809,83.658544,6.466171,3553.889072,2004-05-11 00:16:49.989373,2.717676,7.092455,2003.81509,100.715551
min,10100.000000,6.000000,26.880000,1.000000,482.130000,2003-01-06 00:00:00,1.000000,1.000000,2003.00000,33.000000
25%,10180.000000,27.000000,68.860000,3.000000,2203.430000,2003-11-06 12:00:00,2.000000,4.000000,2003.00000,68.000000
50%,10262.000000,35.000000,95.700000,6.000000,3184.800000,2004-06-15 00:00:00,3.000000,8.000000,2004.00000,99.000000
75%,10333.500000,43.000000,100.000000,9.000000,4508.000000,2004-11-17 12:00:00,4.000000,11.000000,2004.00000,124.000000
max,10425.000000,97.000000,100.000000,18.000000,14082.800000,2005-05-31 00:00:00,4.000000,12.000000,2005.00000,214.000000
std,92.085478,9.741443,20.174277,4.225841,1841.865106,NaN,1.203878,3.656633,0.69967,40.187912


In [112]:
data['status'].value_counts()

status
Shipped       2617
Cancelled       60
Resolved        47
On Hold         44
In Process      41
Disputed        14
Name: count, dtype: int64

In [113]:
(data['status'].value_counts(normalize= True) * 100).round(1)

status
Shipped       92.7
Cancelled      2.1
Resolved       1.7
On Hold        1.6
In Process     1.5
Disputed       0.5
Name: proportion, dtype: float64

above 2 cells
- upper one shows total count of each `ststus` value
- in the lower one it shows percent of each `status` value

In [114]:
data['year_id'].value_counts().sort_index()

year_id
2003    1000
2004    1345
2005     478
Name: count, dtype: int64

In [115]:
data['productline'].value_counts()

productline
Classic Cars        967
Vintage Cars        607
Motorcycles         331
Planes              306
Trucks and Buses    301
Ships               234
Trains               77
Name: count, dtype: int64

In [116]:
data.columns.to_list()

['ordernumber',
 'quantityordered',
 'priceeach',
 'orderlinenumber',
 'sales',
 'orderdate',
 'status',
 'qtr_id',
 'month_id',
 'year_id',
 'productline',
 'msrp',
 'productcode',
 'customername',
 'phone',
 'addressline1',
 'addressline2',
 'city',
 'state',
 'postalcode',
 'country',
 'territory',
 'contactlastname',
 'contactfirstname',
 'dealsize']

In [117]:
data['month_year'] = data['orderdate'].dt.to_period('M')

data[['orderdate','month_year']]

,orderdate,month_year
0,2003-02-24,2003-02
1,2003-05-07,2003-05
2,2003-07-01,2003-07
3,2003-08-25,2003-08
4,2003-10-10,2003-10
...,...,...
2818,2004-12-02,2004-12
2819,2005-01-31,2005-01
2820,2005-03-01,2005-03
2821,2005-03-28,2005-03


In [118]:
month_year = data.groupby('month_year')['sales'].sum().reset_index()
month_year.head(5)

,month_year,sales
0,2003-01,129753.60
1,2003-02,140836.19
2,2003-03,174504.90
3,2003-04,201609.55
4,2003-05,192673.11


**IMPORTANT** — Incomplete Year: The distribution by `month_year` shows that for 2005, only the months from January to May are available. This should be taken into account when making year-over-year comparisons — 2005 is not a “low year”; it is an incomplete year.

## Vizuals

### Sales

In [119]:
fig_sales = px.histogram(data,x= 'sales',nbins= 40,title = 'Distribution of Sales',
                         labels={'SALES': 'Sales Amount (USD)', 'count': 'Frequency (order lines)'},color_discrete_sequence=['#2E86AB'])

fig_sales.update_layout(bargap = 0.1, yaxis_title= 'frequency of orderlines')


fig_sales.show()

In [120]:
data['sales'].describe(include='all').T.round(2)

count     2823.00
mean      3553.89
std       1841.87
min        482.13
25%       2203.43
50%       3184.80
75%       4508.00
max      14082.80
Name: sales, dtype: float64

From the visual above,  we can see taht it is `right skewed`
- **mean** = 3553.89
- **median** = 3184.8

this shows `mean` is higher than `median`, there are rarely higher sales priced orders and it makes `mean` higher

### Year count

In [121]:
year_count = data['year_id'].value_counts().reset_index()
year_count = year_count.sort_values('year_id').reset_index()

In [122]:
fig_year = px.bar(
    year_count,
    x=year_count['year_id'].astype(str),
    y='count',
    title='Orders by Year',
    labels={'x': 'Year', 'count': 'Number of orders'},
    text=year_count['count']
)

fig_year.update_traces(marker_color='#F26419')
fig_year.show()

**Result**  Sales are peaked in `2004` and shortage of `2005` is because of the data is not enough to show full `2005`

### Countries

In [123]:
data.columns.to_list()

['ordernumber',
 'quantityordered',
 'priceeach',
 'orderlinenumber',
 'sales',
 'orderdate',
 'status',
 'qtr_id',
 'month_id',
 'year_id',
 'productline',
 'msrp',
 'productcode',
 'customername',
 'phone',
 'addressline1',
 'addressline2',
 'city',
 'state',
 'postalcode',
 'country',
 'territory',
 'contactlastname',
 'contactfirstname',
 'dealsize',
 'month_year']

In [124]:
countries = data.groupby('country')['sales'].agg('sum').reset_index().sort_values(by='sales',ascending= False)
countries_count = data['country'].value_counts()

country = countries.merge(countries_count,on='country',how='left')
country

,country,sales,count
0,USA,3627982.83,1004
1,Spain,1215686.92,342
2,France,1110916.52,314
3,Australia,630623.10,185
4,UK,478880.46,144
5,Italy,374674.31,113
6,Finland,329581.91,92
7,Norway,307463.70,85
8,Singapore,288488.41,79
9,Denmark,245637.15,63


In [125]:
fig_country = px.bar(
    country.head(5),
    x='country',
    y='sales',
    title='Total Sales by Country',
    labels={'country': 'Country', 'sales': 'Total Sales (USD)'},
    text='sales',
    hover_data=['count']
)
fig_country.update_traces(texttemplate='%{text:.2s}', textposition='inside')
fig_country.update_layout(xaxis_tickangle=-45, template='plotly_white')
fig_country.show()

Among the top 5 countries, it can clearly be seen that the USA stands out significantly, with a value approximately 3 times higher than the second one in terms of `sales counts and also sales total`

In [126]:
top5 = data.groupby('country')['sales'].sum().nlargest(5).index

In [127]:

country_product = data.groupby(['country', 'productline'])['sales'].sum().reset_index()
country_product_top5 = country_product[country_product['country'].isin(top5)]


In [128]:

fig = px.bar(
    country_product_top5,
    x='country',
    y='sales',
    color='productline',
    title='Sales by Product Line — Top 5 Countries',
    labels={'country': 'Country', 'sales': 'Total Sales (USD)', 'productline': 'Product Line'},
    barmode='group',
    category_orders={'country': list(top5)}
)
fig.update_layout(xaxis_tickangle=-45, template='plotly_white')
fig.show()

`Usa` the `Classic cars` are the highest of all products across all countries

in `USA` total sales is `3.6M` and `Classis cars` is `33%`  

### Product

In [129]:
data.columns.to_list()

['ordernumber',
 'quantityordered',
 'priceeach',
 'orderlinenumber',
 'sales',
 'orderdate',
 'status',
 'qtr_id',
 'month_id',
 'year_id',
 'productline',
 'msrp',
 'productcode',
 'customername',
 'phone',
 'addressline1',
 'addressline2',
 'city',
 'state',
 'postalcode',
 'country',
 'territory',
 'contactlastname',
 'contactfirstname',
 'dealsize',
 'month_year']

In [130]:
product = data.groupby('productline')['sales'].sum()

product

productline
Classic Cars        3919615.66
Motorcycles         1166388.34
Planes               975003.57
Ships                714437.13
Trains               226243.47
Trucks and Buses    1127789.84
Vintage Cars        1903150.84
Name: sales, dtype: float64

In [131]:
fig_product = px.bar(
    product.sort_values(ascending=True).reset_index(),
    x='sales',
    y='productline',
    orientation='h',
    title='Total Sales by Product Line',
    labels={'sales': 'Total Sales (USD)', 'productline': 'Product Line'},
    text='sales',
    color='sales',
    color_continuous_scale='Blues'
)
fig_product.update_traces(
    texttemplate='$%{text:,.0f}',
    textposition='inside'
)

fig_product.show()

## RFM Customer segmentetation

In [132]:
s_date = data['orderdate'].max() + pd.Timedelta(days=1)

In [133]:
rfm = (data.groupby('customername').agg(
            recency=('orderdate', lambda x: (s_date - x.max()).days),
            frequency=('ordernumber', 'nunique'),
            monetary=('sales', 'sum')
        ).reset_index())

rfm.head()
               

,customername,recency,frequency,monetary
0,"AV Stores, Co.",196,3,157807.81
1,Alpha Cognac,65,3,70488.44
2,Amica Models & Co.,265,2,94117.26
3,"Anna's Decorations, Ltd",84,4,153996.13
4,Atelier graphique,188,3,24179.96


In [134]:
rfm['recency_rank'] = rfm['recency'].rank(method='first')
rfm['frequency_rank'] = rfm['frequency'].rank(method='first')
rfm['monetary_rank'] = rfm['monetary'].rank(method='first')

rfm['r_score'] = pd.qcut(rfm['recency_rank'], 5, labels=[5, 4, 3, 2, 1]).astype(int)
rfm['f_score'] = pd.qcut(rfm['frequency_rank'], 5, labels=[1, 2, 3, 4, 5]).astype(int)
rfm['m_score'] = pd.qcut(rfm['monetary_rank'], 5, labels=[1, 2, 3, 4, 5]).astype(int)

In [135]:
rfm

,customername,recency,frequency,monetary,recency_rank,frequency_rank,monetary_rank,r_score,f_score,m_score
0,"AV Stores, Co.",196,3,157807.81,53.0,30.0,84.0,3,2,5
1,Alpha Cognac,65,3,70488.44,22.0,31.0,24.0,4,2,2
2,Amica Models & Co.,265,2,94117.26,76.0,2.0,52.0,1,1,3
3,"Anna's Decorations, Ltd",84,4,153996.13,25.0,71.0,83.0,4,4,5
4,Atelier graphique,188,3,24179.96,47.0,32.0,2.0,3,2,1
...,...,...,...,...,...,...,...,...,...,...
87,"Vida Sport, Ltd",275,2,117713.56,78.0,27.0,67.0,1,2,4
88,Vitachrome Inc.,208,3,88041.26,60.0,70.0,48.0,2,4,3
89,"Volvo Model Replicas, Co",194,4,75754.88,51.0,86.0,32.0,3,5,2
90,West Coast Collectables Co.,489,2,46084.64,90.0,28.0,9.0,1,2,1


In [136]:
def assign_segment(row):
    r, f = row['r_score'], row['f_score']
    if r >= 4 and f >= 4:
        return 'Champions'
    elif r >= 3 and f >= 4:
        return 'Loyal Customers'
    elif r >= 4 and  f <= 3:
        return 'Potential Loyalists'
    elif r == 5 and f == 1:
        return 'New Customers'
    elif r >= 3 and f <= 2:
        return 'Promising'
    elif r == 3 and f == 3:
        return 'Need Attention'
    elif r <= 2 and f >= 3:
        return 'At Risk'
    elif r <= 2 and f <= 2:
        return 'Hibernating'
    else:
        return 'Other'

In [137]:
rfm['segment'] = rfm.apply(assign_segment,axis=1)

In [138]:
segment = rfm.groupby('segment').agg(
    customer_count=('customername', 'count'),
    total_revenue=('monetary', 'sum'),
    avg_recency=('recency', 'mean'),
    avg_frequency=('frequency', 'mean')
).round(2).sort_values('total_revenue', ascending=False)


In [139]:
segment

,customer_count,total_revenue,avg_recency,avg_frequency
segment,,,,
Champions,23,4162105.33,55.57,5.30
Hibernating,28,1919521.19,324.25,1.96
Potential Loyalists,14,1125280.59,73.14,2.93
At Risk,9,1098652.31,241.22,3.33
Loyal Customers,7,880787.03,173.86,3.71
Need Attention,7,560020.17,185.00,3.00
Promising,4,286262.23,189.00,3.00


In [140]:
segment['revenue_share_pct'] = (segment['total_revenue'] / rfm['monetary'].sum() * 100).round(1)
segment

,customer_count,total_revenue,avg_recency,avg_frequency,revenue_share_pct
segment,,,,,
Champions,23,4162105.33,55.57,5.30,41.5
Hibernating,28,1919521.19,324.25,1.96,19.1
Potential Loyalists,14,1125280.59,73.14,2.93,11.2
At Risk,9,1098652.31,241.22,3.33,11.0
Loyal Customers,7,880787.03,173.86,3.71,8.8
Need Attention,7,560020.17,185.00,3.00,5.6
Promising,4,286262.23,189.00,3.00,2.9


In [141]:


order = segment.index.tolist()

fig = make_subplots(rows=1, cols=2, subplot_titles=('Customer Count by RFM Segment', 'Total Revenue by RFM Segment'))

fig.add_trace(
    go.Bar(x=order, y=rfm['segment'].value_counts()[order], marker_color='#2E86AB', name='Customer Count'),
    row=1, col=1
)

fig.add_trace(
    go.Bar(x=order, y=segment['total_revenue'], marker_color='#F26419', name='Total Revenue'),
    row=1, col=2
)

fig.update_yaxes(title_text='Number of Customers', row=1, col=1)
fig.update_yaxes(title_text='Total Revenue (USD)', row=1, col=2)
fig.update_xaxes(tickangle=45)

fig.update_layout(showlegend=False, template='plotly_white', height=500, width=1100)
fig.show()

In [142]:
fig = px.scatter(
    rfm,
    x='frequency',
    y='monetary',
    color='segment',
    title='Customer Segments: frequency vs monetary Value',
    labels={'frequency': 'frequency (unique orders)', 'monetary': 'monetary (total sales, USD)'},
    opacity=0.7,
    color_discrete_sequence=px.colors.qualitative.Set2
)

fig.update_traces(marker=dict(size=10))
fig.update_layout(template='plotly_white', legend=dict(x=1.02, y=1, xanchor='left'))
fig.show()

The Champions segment shows us only 23 of 92 customers (41.5%), yet contributes the majority of total revenue

- **Champions:** The most valuable segment — should be retained through loyalty programs and early-access
- **At Risk:** Before gone inactive these customers were frequent buyers — should be targeted with reactivation campaigns (like discount offers)
- **Hibernating:** Long inactive with low frequency — win-back campaigns needed or should be reassessed (high churn probability)
- **New Customers:** Only one order so far, but newly acquired — onboarding and incentives toward a second purchase are important

# Final Conclusion

### Key Findings

| # | Finding |
| - | ------- |
| 1 | Dataset contains 2,823 order-line rows, 307 unique orders, and 92 unique customers |
| 2 | Missing values occur only in 4 address-related columns and are not critical to the core analysis |
| 3 | No duplicate rows found |
| 4 | SALES is right-skewed (mean=$3,553.9, median=$3,184.8) |
| 5 | 2005 is incomplete (Jan–May only); the apparent decline is a data coverage artifact |
| 6 | USA leads with $3.63M in sales, roughly 3× Spain ($1.22M) |
| 7 | Classic Cars is the top product line ($3.92M), representing ~33% of USA sales |
| 8 | Champions generate 41.5% of revenue while representing only 25% of customers |

### RFM Segment Summary

| Segment             | Customers | Total Revenue | Avg Recency | Avg Frequency | Revenue Share |
| ------------------- | --------: | ------------: | ----------: | ------------: | ------------: |
| Champions           |        23 |    $4,162,105 |        55.6 |           5.3 |     **41.5%** |
| Hibernating         |        28 |    $1,919,521 |       324.3 |           2.0 |         19.1% |
| Potential Loyalists |        14 |    $1,125,281 |        73.1 |           2.9 |         11.2% |
| At Risk             |         9 |    $1,098,652 |       241.2 |           3.3 |         11.0% |
| Loyal Customers     |         7 |      $880,787 |       173.9 |           3.7 |          8.8% |
| Need Attention      |         7 |      $560,020 |       185.0 |           3.0 |          5.6% |
| Promising           |         4 |      $286,262 |       189.0 |           3.0 |          2.9% |

### Business Recommendations

* **Champions:** Prioritize retention through loyalty programs, early access, and personalized offers.
* **Hibernating:** Launch win-back campaigns and reassess marketing spend.
* **At Risk:** Target with time-limited reactivation offers.
* **Potential Loyalists / Loyal Customers:** Use upselling and bundles to move customers toward Champion status.
